pkg/schemapatcher/gen.go

```diff
@@ -0,0 +1,355 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schemapatcher
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+
+	"gopkg.in/yaml.v3"
+	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	kyaml "sigs.k8s.io/yaml"
+
+	crdgen "sigs.k8s.io/controller-tools/pkg/crd"
+	crdmarkers "sigs.k8s.io/controller-tools/pkg/crd/markers"
+	"sigs.k8s.io/controller-tools/pkg/genall"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+	yamlop "sigs.k8s.io/controller-tools/pkg/schemapatcher/internal/yaml"
+)
+
+// NB(directxman12): this code is quite fragile, but there are a sufficient
+// number of corner cases that it's hard to decompose into separate tools.
+// When in doubt, ping @sttts.
+//
+// Namely:
+// - It needs to only update existing versions
+// - It needs to make "stable" changes that don't mess with map key ordering
+//   (in order to facilitate validating that no change has occurred)
+// - It needs to collapse identical schema versions into a top-level schema,
+//   if all versions are identical
```

Comment:
this is not special for schemapatcher, but a general requirement for multi-version, schemaful CRDs.

Reply:
I'll make this clearer in the comment -- I was mainly just trying to motivate why "stick the output of `parser.NeedFlattenedSchema` into a single JSON patch operation" is insufficient.